### PR TITLE
explore generators

### PR DESCRIPTION
**Pull request by slavik-lvovsky**

### CODE REVIEW

File: backend/README.md [hunk 1]
```diff
@@ -1,20 +1,64 @@
...
 
-Or use the command line:
-```sh
-# compile server code to out directory:
-npm run compile
-# run the websocket server: 
-npm run ws:run
-```
+## Requirements
+* [node.js](https://www.npmjs.com/package/node) version 10 or higher.
```
Comment (tomer-epstein) [inline backend/README.md:16]: Revert README changes!

File: backend/src/panels/YeomanUIPanel.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+	public loadYeomanUI(uiOptions?: any) {
+		this.disposeWebviewPanel();
+		const webViewPanel = this.createWebviewPanel();
+		this.setWebviewPanel(webViewPanel, uiOptions);
+	}
+
+	public toggleOutput() {
+		this.outputChannel.showOutput();
+	}
```
Comment (tomer-epstein) [inline backend/src/panels/YeomanUIPanel.ts:29]: toggleOutput command only from yeoma-ui webview ?!

File: backend/src/yeomanui.ts [hunk 3]
```diff
@@ -53,9 +53,7 @@ export class YeomanUI {
 
   constructor(rpc: IRpc, youiEvents: YouiEvents, outputChannel: YouiLog, logger: IChildLogger, uiOptions: any, outputPath: string = YeomanUI.PROJECTS) {
     this.rpc = rpc;
-    if (!this.rpc) {
-      throw new Error("rpc must be set");
-    }
```
Comment (tomer-epstein) [inline backend/src/yeomanui.ts:58]: why it is not needed?

File: backend/package.json [hunk 4]
```diff
@@ -89,6 +95,25 @@
...
 					"description": "If chosen, the location of the source code is added to log entries. Warning – this action may slow your extension. We recommend you use it only for debugging.",
 					"scope": "resource"
+				},
+				"Explore Generators.installationLocation": {
+					"type": "string",
+					"markdownDescription": "Install generators in a specified location. If not set, global location is used."
+				},
+				"Explore Generators.autoUpdate": {
+					"type": "boolean",
+					"default": true,
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/package.json:105]: This is what they wanted

File: backend/src/exploregens.ts [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    private readonly INSTALLATION_LOCATION = "Explore Generators.installationLocation";
+    private readonly SEARCH_QUERY = "Explore Generators.searchQuery";
+    private readonly AUTO_UPDATE = "Explore Generators.autoUpdate"
+    private readonly NPM = (process.platform === 'win32' ? 'npm.cmd' : 'npm');
+    private readonly ONE_DAY = 1000 * 60 * 60 * 24;
+
+
+    constructor(context: any, logger: IChildLogger) {
+        this.logger = logger;
+        this.gensBeingHandled = [];
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/exploregens.ts:30]: No this is the only place where it is initialized when an object is created

File: backend/src/exploregens.ts [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        return vscode.workspace.getConfiguration();
+    }
+
+    private async install(gen: any) {
+        const genName: string = gen.package.name;
+        const locationParams = this.getGeneratorsLocationParams();
+        await this.installGenerator(locationParams, genName);
+        const installedGens: string[] = await this.cachedInstalledGeneratorsPromise;
+        installedGens.push(genName);
+        this.cachedInstalledGeneratorsPromise = Promise.resolve(_.uniq(installedGens));
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/exploregens.ts]: Not sure what do you mean, but please recheck this logic have been changed

File: backend/src/exploregens.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    private async uninstall(gen: any) {
+        const genName = gen.package.name;
+        const locationParams = this.getGeneratorsLocationParams();
+        await this.uninstallGenerator(locationParams, genName);
+        const installedGens: string[] = await this.cachedInstalledGeneratorsPromise;
+        this.removeFromArray(installedGens, genName);
+        this.cachedInstalledGeneratorsPromise = Promise.resolve(installedGens);
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/exploregens.ts]: This logic have been changed

File: backend/src/exploregens.ts [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    private showAndLogError(messagePrefix: string, error: any) {
+        const errorMessage = error.toString();
+        this.logger.error(errorMessage);
+        vscode.window.showErrorMessage(`${messagePrefix}: ${errorMessage}`);
+    }
+
+    private getGensQueryURL(query: string, recommended: string) {
+        const api_endpoint = "http://registry.npmjs.com/-/v1/search?text=";
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/exploregens.ts:136]: Done

File: backend/src/exploregens.ts [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+        const actualQuery = encodeURI(`${query} ${recommended}`);
+        return `${api_endpoint}${actualQuery}%20keywords:yeoman-generator%20&size=25&ranking=popularity`;
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/exploregens.ts]: Done

File: backend/src/panels/YeomanUIPanel.ts (hunk 2 above)
Comment (slavik-lvovsky) [reply backend/src/panels/YeomanUIPanel.ts:29]: Yes, they did not defined it in the explore gens mockup

File: backend/src/yeomanui.ts (hunk 3 above)
Comment (slavik-lvovsky) [reply backend/src/yeomanui.ts:58]: rpc is always passed, no need to make this check

File: backend/src/exploregens/exploregens.ts [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    private readonly SEARCH_QUERY = "Explore Generators.searchQuery";
+    private readonly AUTO_UPDATE = "Explore Generators.autoUpdate"
+    private readonly NPM = (process.platform === "win32" ? "npm.cmd" : "npm");
+    private readonly STDOUT = "stdout";
+    private readonly EMPTY = "";
+    private readonly ONE_DAY = 1000 * 60 * 60 * 24;
+
+    constructor(context: any, logger: IChildLogger) {
+        this.logger = logger;
+        this.gensBeingHandled = [];
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/exploregens/exploregens.ts]: this is the only place where it is initialized

File: backend/src/panels/YeomanUIPanel.ts (hunk 2 above)
Comment (tomer-epstein) [reply backend/src/panels/YeomanUIPanel.ts:30]: They might forgot it.

**Pull request merged**